include/networkit/centrality/GroupHarmonicCloseness.hpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * GroupHarmonicCloseness.hpp
+ *
+ * Created on: 15.12.2020
+ *     Author: Eugenio Angriman <angrimae@hu-berlin.de>
+ */
+
+// networkit-format
+
+#ifndef NETWORKIT_CENTRALITY_GROUP_HARMONIC_CLOSENESS_HPP_
+#define NETWORKIT_CENTRALITY_GROUP_HARMONIC_CLOSENESS_HPP_
+
+#include <memory>
+#include <vector>
+
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+namespace GroupHarmonicClosenessDetails {
+template <class>
+class GroupHarmonicClosenessImpl;
+using GroupHarmonicClosenessUnweighted = GroupHarmonicClosenessImpl<count>;
+using GroupHarmonicClosenessWeighted = GroupHarmonicClosenessImpl<edgeweight>;
+} // namespace GroupHarmonicClosenessDetails
+
+/**
+ * @ingroup centrality
+ */
+class GroupHarmonicCloseness final : public Algorithm {
+
+public:
+    /**
+     * Approximation algorithm for the group-harmonic maximization problem. The computed solutions
+     * have a guaranteed $\\lambda(1 - \\frac{1}{2e})$ (directed graphs) and
+     * $\\lambda(1 - \\frac{1}/{e})/2$ (undirected graphs) approximation ratio,
+     * where $\\lambda$ is the ratio
+     * between the minimal and the maximal edge weight. The algorithm is the one proposed in
+     * Angriman et al., ALENEX 2021. The worst-case running time of this approach is quadratic, but
+     * usually much faster in practice.
+     *
+     * @param G The input graph.
+     * @param k Size of the group of nodes.
+     */
+    GroupHarmonicCloseness(const Graph &G, count k = 1);
+
+    ~GroupHarmonicCloseness() override;
+
+    /**
+     * Runs the algorithm.
+     */
+    void run() override;
+
+    /**
+     * Returns the computed group.
+     *
+     * @return The computed group.
+     */
+    const std::vector<node> &groupMaxHarmonicCloseness() const;
+
+    /**
+     * Computes the group-harmonic score of the group of nodes in the given range.
+     *
+     * @param graph The input Graph.
+     * @param first,last The range that contains the vertices in the group.
+     *
+     * @returns The score of the group of nodes in the given range.
+     */
+    template <class InputIt>
+    static double scoreOfGroup(const Graph &graph, InputIt first, InputIt last);
+
+private:
+    const bool weighted;
+    std::unique_ptr<GroupHarmonicClosenessDetails::GroupHarmonicClosenessUnweighted> implUnweighted;
+    std::unique_ptr<GroupHarmonicClosenessDetails::GroupHarmonicClosenessWeighted> implWeighted;
```

Comment:
Two pointers here are kind of unnecessary. We could use a `std::unique_ptr<void *>` (maybe with a comment that it is always one of the two classes, based on the `weighted` variable) and change the code in `run()` to:

```cpp
if (weighted) {
    auto unweightedImpl = static_cast<GroupHarmonicClosenessUnweighted>(impl);
    unweightedImpl->run();
}else{
    auto weightedImpl = static_cast<GroupHarmonicClosenessWeighted>(impl);
    weightedImpl->run();
}
```